nixos/modules/services/hardware/pid-fan-controller.nix

```
@@ -0,0 +1,190 @@
+        description = "Maximum PWM value";
+      };
+      cutoff = lib.mkOption {
+        default = false;
+        type = lib.types.bool;
+        description = "Whether to stop the fan when `minPwm` is reached";
+      };
+      heatPressureSrcs = lib.mkOption {
+        type = lib.types.nonEmptyListOf lib.types.str;
+        description = "Heat pressure sources affected by the fan";
+      };
+    };
+  };
+in
+{
+  options.services.pid-fan-controller = {
+    enable = lib.mkEnableOption ''
+      the PID fan controller, which controls the configured fans by running
+      a closed-loop PID control loop.
+    '';
+    package = lib.mkPackageOption pkgs "pid-fan-controller" { };
+    settings = {
+      interval = lib.mkOption {
+        default = 500;
+        type = lib.types.int;
+        description = "Interval between controller cycles in ms";
+      };
+      heatSrcs = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule heatSource);
+        description = "List of heat sources to be monitored";
+        example = ''
+          [
+            {
+              name = "cpu";
+              wildcardPath = "/sys/devices/pci0000:00/0000:00:18.3/hwmon/hwmon*/temp1_input";
+              pidParams = {
+                setPoint = 60;
+                P = -5.0e-3;
+                I = -2.0e-3;
+                D = -6.0e-3;
+              };
+            }
+          ];
+        '';
+      };
+      fans = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule fan);
+        description = "List of fans to be controlled";
+        example = ''
+          [
+            {
+              wildcardPath = "/sys/devices/platform/nct6775.2592/hwmon/hwmon*/pwm1";
+              minPwm = 60;
+              maxPwm = 255;
+              heatPressureSrcs = [
+                "cpu"
+                "gpu"
+              ];
+            }
+          ];
+        '';
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    #map camel cased attrs into snake case for config
+    environment.etc."pid-fan-settings.json".text = builtins.toJSON {
+      interval = cfg.settings.interval;
+      heat_srcs = map (heatSrc: {
+        name = heatSrc.name;
+        wildcard_path = heatSrc.wildcardPath;
+        PID_params = {
+          set_point = heatSrc.pidParams.setPoint;
+          P = heatSrc.pidParams.P;
+          I = heatSrc.pidParams.I;
+          D = heatSrc.pidParams.D;
+        };
+      }) cfg.settings.heatSrcs;
+      fans = map (fan: {
+        wildcard_path = fan.wildcardPath;
+        min_pwm = fan.minPwm;
+        max_pwm = fan.maxPwm;
+        cutoff = fan.cutoff;
+        heat_pressure_srcs = fan.heatPressureSrcs;
+      }) cfg.settings.fans;
+    };
+
+    systemd.services.pid-fan-controller = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = [ (lib.meta.getExe cfg.package) ];
+        ExecStopPost = [ "${lib.meta.getExe cfg.package} disable" ];
```

```suggestion
        ExecStart = [ (lib.getExe cfg.package) ];
        ExecStopPost = [ "${lib.getExe cfg.package} disable" ];
```
For consistency. You haven't declared the `lib` category elsewhere.